api/transport/invoker.go

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package transport
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"runtime/debug"
+	"time"
+
+	"go.uber.org/yarpc/yarpcerrors"
+	"go.uber.org/zap"
+)
+
+// UnaryInvokeRequest encapsulates minimum arguments to invoke a unary handler
+type UnaryInvokeRequest struct {
+	Context        context.Context
+	StartTime      time.Time
+	Request        *Request
+	ResponseWriter ResponseWriter
+	Handler        UnaryHandler
+	Options        *InvokerOptions
```

Comment:
It looks like we're mixing parameter struct and functional options here.

Usually, we should use one of those per API since they're both just ways of
accepting new optional arguments in the future.

With this many "required" arguments, parameter structs are usually the way to
go, at which point their optional fields (current and new) can just have a
comment.

So consider something like the following,

```
type UnaryInvokeRequest struct {
  Context context.Context
  StartTime time.Time
  ...

  Logger *zap.Logger  // optional
}
```

And in the future, we can add metrics there as optional too.

This has the side-effect of having duplicate fields across these Request
structs and making the call to handlePanic a bit noisy, but it's a reasonable
tradeoff IMO.

Reply:
re-duping